### PR TITLE
Meta files standarization

### PR DESCRIPTION
**Pull request by miquelmiro3**
Please guys, with this pr the resource manager changes a lot with all the new Meta file structure. If you can don't just review the code make some QA and test if there is something that doesn't work as it supouse to. Thanks!

Test cases:
- Delete library
- Import of new assets
- Change textures in materials
- Create scenes

Metas should be forever when you include an asset its meta should be also uploaded, so the relation UID asset is never lost. Don't delete metas, if using saved scenes.

Please if there any bug send us the steps to reproduce it. Thanks!

### CODE REVIEW

**Comment by VBoschMatas** [comment]
Everything looks fine, but if I double click on any .meta file it will try to load a model from it. But for now we can just literally avoid doing that directly, and after that at some point check the meta type and if we have an object selected and we double click a material change its meterial to the selected and for niche or confusing things just do nothing when double clicking.

**Review by VBoschMatas — APPROVED** [review]

**Comment by alvarosoppelsa** [comment]
I'll fix the bug right now and I'll try to add the material change for the selected object. Thank you for testing!

File: Source/src/modules/ModuleResources.h [hunk 1]
```diff
@@ -21,36 +21,36 @@ namespace Hachiko
...
-
+        void AssetsLibraryCheck();
+        void HandleResource(const std::filesystem::path& path);
 
     private:
+        // TODO: Use only loaded_resources map
         std::map<std::string, ResourceModel*> models;
         std::map<std::string, ResourceMaterial*> materials;
         std::map<std::string, ResourceTexture*> textures;
         std::map<UID, ResourceMesh*> meshes;
```
Comment (Bernatmago) [inline Source/src/modules/ModuleResources.h:35]: we can remove these specific maps

**Review by Bernatmago — CHANGES_REQUESTED** [review]
I talked with Miquel and we should not regenerate meta files because timestamp changes and git wants to track the changes (or remove timestamp from them otherwise)

**Comment by alvarosoppelsa** [comment]
@VBoschMatas I've added material selection by double clicking in WindowResources. Could you give it a look? Let me know what you think. Thanks! :)

**Comment by VBoschMatas** [comment]
> @VBoschMatas I've added material selection by double clicking in WindowResources. Could you give it a look? Let me know what you think. Thanks! :)

Looks nice and intuitive, I've seen that double clicking on other things does not crash the engine. Good job!

**Review by armandojaga — CHANGES_REQUESTED** [review]

File: Source/src/importers/Importer.h [hunk 2]
```diff
@@ -31,9 +31,18 @@ namespace Hachiko
             return type;
         }
 
-        virtual void Import(const char* path) = 0;
+        virtual void Import(const char* path, YAML::Node& meta) = 0;
+        virtual void ImportWithMeta(const char* path, YAML::Node& meta) = 0;
```
[the remark this answers is not in the document]
Comment (alvarosoppelsa) [reply Source/src/importers/Importer.h:35]: It's a really good point. We've splitted functionality when we are importing a new file and when already have created the meta file. We've talk to do a code improvement, is it ok for you if we add a TODO comment for reviewing this after merge?

File: Source/src/importers/ModelImporter.h [hunk 3]
```diff
@@ -13,11 +13,14 @@ namespace Hachiko
...
+        void Import(const char* path, YAML::Node& meta) override;
+        void ImportWithMeta(const char* path, YAML::Node& meta) override;
         void Save(const Resource* resource) override;
-        Resource* Load(const char* model_path) override;
+        Resource* Load(UID id) override;
+        void Delete(const YAML::Node& meta) override;
+
         [[nodiscard]] bool IsImported(const char* path) override;
-        Resource* CherryImport(int mesh_index, const UID uid, const char* model_path);
+        //Resource* CherryImport(int mesh_index, const UID uid, const char* model_path); // TODO: To remove
```
[the remark this answers is not in the document]
Comment (alvarosoppelsa) [reply Source/src/importers/ModelImporter.h]: Done

File: Source/src/importers/ModelImporter.cpp [hunk 4]
```diff
@@ -176,58 +195,69 @@ void Hachiko::ModelImporter::LoadChildren(YAML::Node& node, YAML::Node& meshes,
 
 void Hachiko::ModelImporter::Save(const Resource* resource) {}
 
-Hachiko::Resource* Hachiko::ModelImporter::CherryImport(int mesh_index, const UID uid, const char* model_path)
+//Hachiko::Resource* Hachiko::ModelImporter::CherryImport(int mesh_index, const UID uid, const char* model_path)
```
[the remark this answers is not in the document]
Comment (alvarosoppelsa) [reply Source/src/importers/ModelImporter.cpp]: Done

**Review by Bernatmago — APPROVED** [review]

File: Source/src/importers/AnimationImporter.cpp [hunk 5]
```diff
@@ -7,11 +7,16 @@
 
 Hachiko::AnimationImporter::AnimationImporter() : Importer(Importer::Type::ANIMATION) {}
 
-void Hachiko::AnimationImporter::Import(const char* path) {}
+void Hachiko::AnimationImporter::Import(const char* path, YAML::Node& meta) {}
+
+void Hachiko::AnimationImporter::ImportWithMeta(const char* path, YAML::Node& meta)
```
[the remark this answers is not in the document]
Comment (miquelmiro3) [reply Source/src/importers/AnimationImporter.cpp:12]: The main difference is that in the first one the import puts data in the meta and uses only the id that contains. In the ImportWithMeta it uses all the data inside the meta to make the import, and writes only the new data of the assets. For example, on a model import the first one would write on the meta that the mesh1 has the id 1, and in the second one the importwithmeta would read that the mesh1 has the id 1 and would import it acording that data.
Hope this makes sense.

File: .gitignore [hunk 6]
```diff
@@ -363,16 +364,11 @@ Source/.idea/
 # ImGui
 imgui.ini
 
-# Wwise
-!Source/vendors/wwise/x64_vc160/Debug/
-!Source/vendors/wwise/include/AK/Tools/Win32/
-
```
Comment (baransrc) [inline .gitignore:369]: Why are we deleting these?

File: .gitignore [hunk 7]
```diff
@@ -363,16 +364,11 @@ Source/.idea/
...
 # Hachiko Engine
 Source/VTune Profiler Results/
-Source/.idea/
-
```
Comment (baransrc) [inline .gitignore:373]: Same here

**Review by baransrc — COMMENTED** [review]
Good work, there are some things need to be clarified. Also Github didn't let me review everything, I think there is an issue with it right now. I could only submit my comments from 2 days ago + .gitignore.

File: Source/src/importers/Importer.h (hunk 2 above)
[the remark this answers is not in the document]
Comment (Bernatmago) [reply Source/src/importers/Importer.h:46]: Import with meta should be load from meta and there will be no confusion <3

[the remark this answers is not in the document]
Comment (Bernatmago) [reply Source/src/importers/Importer.h:46]: You can also change import to Import and fill/populate meta

File: Source/src/core/serialization/Definitions.h [hunk 8]
```diff
@@ -84,30 +86,12 @@
...
-#define VIDEO_LIBRARY "video_library"
-#define SCRIPTS_LIBRARY "scripts_library"
-#define MATERIALS_LIBRARY "materials_library"
-#define SHADERS_LIBRARY "shaders_library"
-#define ANIMATIONS_LIBRARY "animations_library"
-#define SKYBOX_LIBRARY "skybox_library"
+#define LIBRARY_COMMON_NAME "_library"
 
 // Render
 #define RENDER_NODE "render"
```
[the remark this answers is not in the document]
Comment (Bernatmago) [reply Source/src/core/serialization/Definitions.h:97]: Yep could be better

File: Source/src/core/preferences/src/ResourcesPreferences.h [hunk 9]
```diff
@@ -19,30 +19,38 @@ namespace Hachiko
...
+            {Resource::Type::MESH, "assets/meshes/"},
+            {Resource::Type::TEXTURE, "assets/textures/"},
+            {Resource::Type::AUDIO, "assets/audio/"},
+            {Resource::Type::VIDEO, "assets/video/"},
+            {Resource::Type::SCRIPT, "assets/scripts/"},
+            {Resource::Type::MATERIAL, "assets/materials/"},
+            {Resource::Type::SHADER, "assets/shaders/"},
+            {Resource::Type::ANIMATION, "assets/animations/"},
+            {Resource::Type::SKYBOX, "assets/skybox/"},
+            {Resource::Type::FONT, "library/fonts/"},
```
Comment (baransrc) [inline Source/src/core/preferences/src/ResourcesPreferences.h:34]:
```suggestion
            {Resource::Type::FONT, "assets/fonts/"},
```

**Review by baransrc — APPROVED** [review]

**Pull request merged**